src/cuda/api/array.hpp

```diff
@@ -0,0 +1,153 @@
+/**
+ * @file array.hpp
+ * 
+ * @brief A CUDA array wrapper class. Especially useful
+ * if used in combination with CUDA textures.
+ *
+ */
+
+#pragma once
+#ifndef CUDA_API_WRAPPERS_ARRAY_HPP
+#define CUDA_API_WRAPPERS_ARRAY_HPP
+
+#include <cuda/api/current_device.hpp>
+#include <cuda/api/error.hpp>
+
+#include <cuda_runtime.h>
+#include <cuda_runtime_api.h>
+
+#include <algorithm>
+#include <array>
+#include <numeric>
+
+namespace cuda {
+
+template<bool AssumedCurrent> class device_t;
+
+namespace array {
+
+namespace detail {
+
+/**
+ * @brief Do not use directly
+ *
+ */
+template <typename T, size_t NDIMS>
+class array_base {
+
+protected:
+	array_base(dimensions_t<NDIMS> dims, cudaArray* array_ptr) :
+	    dims_(dims),
+	    array_ptr_(array_ptr) {}
+
+public:
+	array_base()                        = delete;
+	array_base(const array_base& other) = delete;
+
+	array_base(array_base&& other) :
+	    array_ptr_(other.array_ptr_),
+	    dims_(other.dims_) {
+		other.array_ptr_ = nullptr;
+		other.dims_      = {0, 0, 0};
+	}
+
+	~array_base() {
+		if (array_ptr_) {
+			auto status = cudaFreeArray(array_ptr_);
+			throw_if_error(status, "failed freeing CUDA array");
+		}
+	}
+
+	cudaArray* get() const noexcept { return array_ptr_; }
+
+	dimensions_t<NDIMS> dims() const noexcept { return dims_; }
+
+	size_t size() const noexcept {
+		using dimensions_value_type = typename dimensions_t<NDIMS>::value_type;
+		return std::accumulate(
+		    dims_.begin(), dims_.end(), static_cast<dimensions_value_type>(1),
+		    [](const dimensions_value_type& a, const dimensions_value_type& b) { return a * b; });
+	}
+
+	size_t size_bytes() const noexcept { return size() * sizeof(T); }
+
+protected:
+	dimensions_t<NDIMS> dims_;
+	cudaArray*          array_ptr_;
+};
+
+} // namespace detail
+
+/**
+ * @brief Wrapper for 2D and 3D arrays
+ * 
+ * Useful in combination with texture memory.
+ * 
+ * @note Only template specializations of this class can be created, because
+ * CUDA does not support any number of dimensions for a CUDA array.
+ */
+template <typename T, size_t NDIMS>
+class array_t {
+public:
+	array_t() = delete;
+};
+
+template <typename T>
+class array_t<T, 3> : public detail::array_base<T, 3> {
+
+private:
```

Comment:
I could not think of a reasonable example when this can do **anything** good. If a child class wants to use the `malloc_3d_cuda_array` function directly, it is obvious that the `array_t` class is not a good wrapper class.

Reply:
@codecircuit : The fact that we can't think of an example is not proof that it doesn't exist. Here's one for you: A subclass which makes a lot of logging calls; or checks some system state before making any CUDA runtime API calls. etc.

Reply:
But why should such a class **inherit** from this array class and want to use that `malloc_3d_cuda_array` function directly? For me that is a case that sounds really wrong.

Reply:
I still do not see any reasonable case to make this static function protected.

- A subclass which wants to make some *actions* before any CUDA runtime API calls, but inherits from this array class. Why should such a class want to access `malloc_3d_cuda_array` directly? Sounds not reasonable to me.

- A class which wants to simulate a CUDA device and inherits from the array class, e.g. `simulated_cuda_array`: should in fact overwrite `malloc_3d_cuda_array`, which is not possible, because it is a static member function of the base class.

But maybe it is a good idea to give another template argument to the array class, which is the allocator type. Similar to `std::vector`. What you think about that idea?

Reply:
@codecircuit : 

1. I don't know why. We're writing API wrappers. It's not for us to forbid users from doing that. We ask the question the other way around: Is there an important reason for this method to be `private` rather than `protected`? Is it important for subclasses _not_ to be able to access it?

2. If we though the array class should not be inherited from we should have been marking it `final`. (And we aren't.)